### PR TITLE
Feat/20/initial structure for generating recommendations

### LINKED ISSUE #20: Create initial structure for generating recommendations

### PR DESCRIPTION
**Pull request by tklausen**
closes #20

### CODE REVIEW

File: src/main/java/org/fhirvp/model/mapper/ImmunizationRecommendationMapped.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    public String getId() {
+        return _raw.getId();
+    }
+
+    public String getPatientId() {
+        String[] referenceParts = _raw.getPatient().getReference().getValue().split("/");
+        return referenceParts[referenceParts.length - 1];
+    }
+
+    public Calendar getCreatedDate() {
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/model/mapper/ImmunizationRecommendationMapped.java]: Why are we using `Calendar` here and not `LocalDate`? https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalDate.html

File: src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                        .build())
+                .build();
+    }
+
+    public ImmunizationRecommendation buildImmunizationRecommendation
+            (
+                    String patientId, String vaccineCode, String targetDiseaseCode, ForecastStatus forecastStatus,
+                    Optional<Calendar> dateToGive, Optional<Calendar> earliestDateToGive, Optional<Calendar> latestDateToGive,
+                    List<String> supportingImmunizationIds, boolean isDeactivated, Optional<String> fulfillingImmunizationId,
+                    String supportingPopulationRecommendationId, String recommendedVaccinationDoseId
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java]: This formatting style is not readable at all for me.

File: src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@ApplicationScoped
+public class FHIRResourceBuilderUseCaseImpl implements FHIRResourceBuilderUseCase {
+
+    @Inject
+    FHIRClientConfig fhirClientConfig;
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java]: You don't need this, the FHIR server accepts relative URLs without the base URL.

File: src/main/java/org/fhirvp/usecase/NotificationUseCase.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.fhirvp.usecase;
+
+import org.fhirvp.model.FHIRNotificationEvent;
+import org.fhirvp.ports.impl.fhir.exception.FHIRServerException;
+
+public interface NotificationUseCase {
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/usecase/NotificationUseCase.java]: Naming suggestion: `RouteNotificationUseCase` with method `route(FHIRNotification notification)`

File: src/main/java/org/fhirvp/usecase/FHIRResourceBuilderUseCase.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.ibm.fhir.model.resource.Basic;
+import com.ibm.fhir.model.resource.ImmunizationRecommendation;
+import org.fhirvp.model.ForecastStatus;
+
+import java.util.Calendar;
+import java.util.List;
+import java.util.Optional;
+
+public interface FHIRResourceBuilderUseCase {
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/usecase/FHIRResourceBuilderUseCase.java]: I don't think this belongs here, it should be in the model package.

File: src/main/java/org/fhirvp/usecase/impl/PatientNotification.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.fhirvp.model.FHIRNotificationEvent;
+import org.fhirvp.ports.ActiveVaccinationSchemePort;
+import org.fhirvp.ports.PatientPort;
+import org.fhirvp.ports.impl.fhir.exception.FHIRServerException;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+@ApplicationScoped
+public class PatientNotification {
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/usecase/impl/PatientNotification.java]: Naming suggestion: `OnPatientNotificationUseCase`. The current name sounds like a model class to me.

Comment (n-krueger) [reply src/main/java/org/fhirvp/usecase/impl/PatientNotification.java]: This UseCase should also have an interface definition to decouple the implementation from the classes that use this UseCase.

**Review by n-krueger — CHANGES_REQUESTED** [review]
I like the general structure, but I left some comments on the details.

File: src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                                .code(Code.builder()
+                                        .value(loincCode)
+                                        .build())
+                                .build())
+                        .build())
+                .value(DateTime.of(LocalDate.of(date.get(Calendar.YEAR), date.get(Calendar.MONTH), date.get(Calendar.DAY_OF_MONTH))))
+                .build();
+    }
+
+    private Reference buildSupportingImmunization(String id) {
```
Comment (n-krueger) [inline src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java]: You are not using the `id` parameter in the method body.

**Review by n-krueger — CHANGES_REQUESTED** [review]

File: src/main/java/org/fhirvp/model/mapper/ImmunizationRecommendationMapped.java (hunk 1 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/model/mapper/ImmunizationRecommendationMapped.java]: Addressed at c5ce4bf

File: src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java (hunk 2 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java]: Addressed at a9b6f4d

File: src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java (hunk 3 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java]: Addressed at a9b6f4d

File: src/main/java/org/fhirvp/usecase/FHIRResourceBuilderUseCase.java (hunk 5 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/usecase/FHIRResourceBuilderUseCase.java]: Addressed at a9b6f4d

File: src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java (hunk 7 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/usecase/impl/builder/FHIRResourceBuilderUseCaseImpl.java]: Addressed at a9b6f4d

File: src/main/java/org/fhirvp/usecase/NotificationUseCase.java (hunk 4 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/usecase/NotificationUseCase.java]: Addressed at 86512c8

File: src/main/java/org/fhirvp/usecase/impl/PatientNotification.java (hunk 6 above)
Comment (tklausen) [reply src/main/java/org/fhirvp/usecase/impl/PatientNotification.java]: Addressed at 86512c8

**Review by n-krueger — APPROVED** [review]
LGTM!

**Pull request merged**